[CIR] Add Function Argument Demotion support

adams381 opened the pull request:
This PR migrates the Function Argument Demotion feature from the incubator repository to upstream. The feature handles K&R-style function parameters that are promoted (e.g., short->int, float->double) and demotes them back to their declared types.

## Changes
- Add emitArgumentDemotion helper function for type demotion
- Create emitFunctionProlog function to handle function prologue setup (addresses existing TODO to move parameter handling logic)
- Move parameter handling logic into emitFunctionProlog
- Add test case kr-func-promote.c to verify the feature

Tested: All CIR tests pass (320/321, 99.69%). The one unsupported test is an expected failure.

adams381 committed:
Add Function Argument Demotion support

This commit migrates the Function Argument Demotion feature from
the incubator repository to upstream. The feature handles K&R-style
function parameters that are promoted (e.g., short->int, float->double)
and demotes them back to their declared types.

Changes:
- Add emitArgumentDemotion helper function for type demotion
- Create emitFunctionProlog function to handle function prologue setup
  (addresses existing TODO to move parameter handling logic)
- Move parameter handling logic into emitFunctionProlog
- Add test case kr-func-promote.c to verify the feature

Tested: All CIR tests pass (320/321, 99.69%). The one unsupported
test is an expected failure.

adams381 committed:
Fix null pointer dereference in emitFunctionProlog

The emitFunctionProlog function was dereferencing fd->getBody()
unconditionally without checking if fd is null or if getBody()
returns null. Similarly, startFunction was dereferencing
fd->getBody()->getEndLoc() without null checks.

Fix by:
- Adding SourceLocation bodyBeginLoc parameter to emitFunctionProlog
- Computing bodyBeginLoc safely in startFunction with null checks
- Computing bodyEndLoc safely in startFunction with null checks
- Using fallback locations (startLoc/loc) when fd is null or has no body

This prevents crashes when fd is null (e.g., for non-function
declarations) or when the function has no body (declaration only).

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.cpp:
```diff
@@ -412,38 +412,31 @@ void CIRGenFunction::LexicalScope::emitImplicitReturn() {
-  mlir::Block *entryBB = &fn.getBlocks().front();
-  builder.setInsertionPointToStart(entryBB);
+  return cgf.getBuilder().CIRBaseBuilderTy::createCast(cir::CastKind::floating,
+                                                       value, ty);
```
```suggestion
  return cgf.getBuilder().createFloatingCast(value, ty);
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.cpp:
```diff
@@ -412,38 +412,31 @@ void CIRGenFunction::LexicalScope::emitImplicitReturn() {
+                                        mlir::Block *entryBB,
+                                        const FunctionDecl *fd,
+                                        SourceLocation bodyBeginLoc) {
   // Declare all the function arguments in the symbol table.
```
Classic codegen starts this function by checking the FunctionDecl for `NakedAttr` and returning immediately if it's present. We should have a check for that.

```
  // Naked functions don't have prologues.
  if (fd && fd->hasAttr<NakedAttr>()) {
    cgm.errorNYI(bodyBeginLoc, "naked function decl");
  }
```

andykaylor reviewed (commented):
This looks good. I have just a couple of nits and a test request.

adams381 committed:
Address reviewer feedback for Function Argument Demotion PR

- Use createFloatingCast instead of createCast with CastKind::floating
- Add NakedAttr check in emitFunctionProlog to handle naked functions early
- Add LLVM and OGCG checks to kr-func-promote.c test

Addresses review comments from PR #170915

andykaylor reviewed (approved):
lgtm

adams381 committed:
Fix formatting: wrap long comment line in CIRGenFunction.h

adams381 committed:
Fix test: make pattern flexible to handle optional no_inline attribute

The no_inline attribute is added at -O0 (default), so the test pattern
should be flexible to match with or without it.
